src/foam/nanos/rope/ROPE.js

```diff
@@ -0,0 +1,221 @@
+            break;
+          default: 
+        }
+        return sourceObjs;
+      `
+    },
+    {
+      name: 'retrieveProperty',
+      args: [
+        { name: 'obj', javaType: 'FObject' },
+        { name: 'objClass', javaType: 'Class' },
+        { name: 'prefix', javaType: 'String' },
+        { name: 'propertyName', javaType: 'String' },
+        { name: 'x', javaType: 'X...' }
+      ],
+      javaType: '<T> T',
+      documentation: `
+      This function returns the value of a property on a object.
+      args: 
+          prefix - the prefix of the property's getter method. i.e., "get", "find"
+          propertyName - the name of the property to retrieve the value of 
+          x - optional argument, provide if the property's getter takes the context as an argument
+      `,
+      javaCode: `
+        if ( objClass == null ) objClass = obj.getClass();
+        Method method;
+        try {
+          Class<?>[] params = ( x.length > 0 ) ? new Class<?>[]{X.class} : new Class<?>[]{};
+          method = objClass.getDeclaredMethod(
+            prefix + propertyName.substring(0, 1).toUpperCase() + propertyName.substring(1),
+            params
+          );
+          method.setAccessible(true);
+          T ret = x.length > 0 ? (T) method.invoke((FObject) obj, x[0]) : (T) method.invoke((FObject) obj);
+          return ret;
+        } catch ( NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e ) {
+          if ( e instanceof NoSuchMethodException && ((Class) objClass).getSuperclass() != null ) return retrieveProperty(obj, ((Class) objClass).getSuperclass(), prefix, propertyName, x);
+          System.err.println("ROPE ERROR: Attempted access on non-existant property ");
+          e.printStackTrace();
+        } 
+        return null;
+      `
+    },
+    {
+      name: 'getNextRelationships',
+      args: [
+        { name: 'relationshipKey', javaType: 'String' },
+        { name: 'crudKey', javaType: 'CrudOperation' },
+        { name: 'propertyKey', javaType: 'String' }
+      ],
+      javaType: 'List<String>',
+      documentation: `
+      This function returns a list of relationshipKeys to filter the ropeDAO within intermediate steps of the ropeSearch
+      args :
+        relationshipKey - if the relationshipKey is specified, get list of keys in relationshipMap[relationshipKey]
+        crudKey - if the crudKey is specified, get list of keys in crudMap[crudKey]["__default__"], or
+        propertyKey - if the propertyKey is specified, get list of keys in crudMap[crudKey][propertyKey] or crudMap[crudKey]["__default__"], if the first returns null
+      `,
+      javaCode: `
+        List<String> next = new ArrayList<String>();
+    
+        if ( crudKey != null ) {
+          next = getCrudMap() == null ? null : getCrudMap().get(crudKey, propertyKey);
+        } else if ( relationshipKey != null && ! relationshipKey.equals("") ) {
```

Comment:
`relationshipKey.equals("")` should be changed to `"".equals(relationshipKey)` (this is recommended in our code review guide to prevent NPEs)